src/lib.rs

```diff
@@ -0,0 +1,379 @@
+// derivative works of, display, perform, and distribute the Software and make,
+// use, sell, offer for sale, import, export, have made, and have sold the
+// Software and the Larger Work(s), and to sublicense the foregoing rights on
+// either these or other terms.
+//
+// This license is subject to the following condition: The above copyright
+// notice and either this complete permission notice or at a minimum a reference
+// to the UPL must be included in all copies or substantial portions of the
+// Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+
+#![crate_name = "gcmalloc"]
+#![crate_type = "rlib"]
+#![feature(core_intrinsics)]
+
+// Allocators are not allowed to depend on the standard library which in turn
+// requires an allocator in order to avoid circular dependencies. This crate,
+// however, can use all of libcore.
+#![feature(no_std)]
+#![no_std]
+
+// Our system allocator will use the in-tree libc crate for FFI bindings. Note
+// that currently the external (crates.io) libc cannot be used because it links
+// to the standard library (e.g. `#![no_std]` isn't stable yet), so that's why
+// this specifically requires the in-tree version.
+
+extern crate libc;
+
+static SIZE_ALLOC_TABLE: usize = (1024 * 1024 * 1024) * 2; // 2KB
+
+/// A contiguous chunk of memory which records metadata about each pointer
+/// allocated by the allocator. Any pointer can be queried during runtime to
+/// determine whether it points directly to, or inside an allocation block on
+/// the Rust heap.
+///     
+/// The size of the allocation block is also recorded, so for each pointer into
+/// the heap the exact size of the allocation block can be known, allowing a
+/// conservative GC to know which additional machine words must be scanned.
+///
+/// TODO: Grow the AllocList if it exceeds its size.
+struct AllocList {
+    start: *const Block,
+    next_free: usize,
+    can_bump: bool,
+}
+
+struct AllocListIter<'a> {
+    alloc_list: &'a AllocList,
+    index: usize,
+}
+
+struct AllocListIterMut<'a> {
+    alloc_list: &'a mut AllocList,
+    index: usize,
+}
+
+impl<'a> Iterator for AllocListIterMut<'a> {
+    type Item = &'a mut Block;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        // It's UB to call `.add` on a pointer past its allocation bounds, so we
+        // need to check that it's within range before turning it into a pointer
+        // and dereferencing it
+        let ulimit = self.alloc_list.start as usize + SIZE_ALLOC_TABLE;
```

Comment:
Might this check be easier as something like:

```rust
if self.index * size_of::<...>() >= SIZE_ALLOC_TABLE {
    None
} else {
    unsafe { Some(...)}
}
```

This way you remove the (very small chance) that you overflow `ptr`.

Reply:
That's much nicer! It would need to be `self.index + 1 * ...` as it's 0-indexed.